File: dotnet/src/Agents/AzureAI/RunPollingOptions.cs
<<<<<<< HEAD
﻿// Copyright (c) Microsoft. All rights reserved.
=======
// Copyright (c) Microsoft. All rights reserved.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
using System;

namespace Microsoft.SemanticKernel.Agents.AzureAI;

/// <summary>
/// Configuration and defaults associated with polling behavior for Assistant API run processing.
/// </summary>
public sealed class RunPollingOptions
{
    /// <summary>
<<<<<<< HEAD
    /// The default maximum number or retries when monitoring thread-run status.
=======
    /// Gets the default maximum number of retries when monitoring thread-run status.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public static int DefaultMaximumRetryCount { get; } = 3;

    /// <summary>
<<<<<<< HEAD
    /// The default polling interval when monitoring thread-run status.
=======
    /// Gets the default polling interval when monitoring thread-run status.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public static TimeSpan DefaultPollingInterval { get; } = TimeSpan.FromMilliseconds(500);

    /// <summary>
<<<<<<< HEAD
    /// The default back-off interval when  monitoring thread-run status.
=======
    /// Gets the default back-off interval when monitoring thread-run status.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public static TimeSpan DefaultPollingBackoff { get; } = TimeSpan.FromSeconds(1);

    /// <summary>
<<<<<<< HEAD
    /// The default number of polling iterations before using <see cref="RunPollingBackoff"/>.
=======
    /// Gets the default number of polling iterations before using <see cref="RunPollingBackoff"/>.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public static int DefaultPollingBackoffThreshold { get; } = 2;

    /// <summary>
<<<<<<< HEAD
    /// The default polling delay when retrying message retrieval due to a 404/NotFound from synchronization lag.
=======
    /// Gets the default polling delay when retrying message retrieval due to a 404/NotFound from synchronization lag.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public static TimeSpan DefaultMessageSynchronizationDelay { get; } = TimeSpan.FromMilliseconds(500);

    /// <summary>
<<<<<<< HEAD
    /// The maximum retry count when polling thread-run status.
    /// </summary>
    /// <remarks>
    /// Only affects failures that have the potential to be transient.  Explicit server error responses
    /// will result in immediate failure.
=======
    /// Gets or sets the maximum retry count when polling thread-run status.
    /// </summary>
    /// <remarks>
    /// This value only affects failures that have the potential to be transient.
    /// Explicit server error responses will result in immediate failure.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </remarks>
    public int MaximumRetryCount { get; set; } = DefaultMaximumRetryCount;

    /// <summary>
<<<<<<< HEAD
    /// The polling interval when monitoring thread-run status.
=======
    /// Gets or sets the polling interval when monitoring thread-run status.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public TimeSpan RunPollingInterval { get; set; } = DefaultPollingInterval;

    /// <summary>
<<<<<<< HEAD
    /// The back-off interval when  monitoring thread-run status.
=======
    /// Gets or sets the back-off interval when monitoring thread-run status.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public TimeSpan RunPollingBackoff { get; set; } = DefaultPollingBackoff;

    /// <summary>
<<<<<<< HEAD
    /// The number of polling iterations before using <see cref="RunPollingBackoff"/>.
=======
    /// Gets or sets the number of polling iterations before using <see cref="RunPollingBackoff"/>.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public int RunPollingBackoffThreshold { get; set; } = DefaultPollingBackoffThreshold;

    /// <summary>
<<<<<<< HEAD
    /// The polling delay when retrying message retrieval due to a 404/NotFound from synchronization lag.
=======
    /// Gets or sets the polling delay when retrying message retrieval due to a 404/NotFound from synchronization lag.
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    /// </summary>
    public TimeSpan MessageSynchronizationDelay { get; set; } = DefaultMessageSynchronizationDelay;

    /// <summary>
    /// Gets the polling interval for the specified iteration count.
    /// </summary>
<<<<<<< HEAD
    /// <param name="iterationCount">The number of polling iterations already attempted</param>
=======
    /// <param name="iterationCount">The number of polling iterations already attempted.</param>
>>>>>>> 6829cc1483570aacfbb75d1065c9f2de96c1d77e
    public TimeSpan GetPollingInterval(int iterationCount)
    {
        return iterationCount > this.RunPollingBackoffThreshold ? this.RunPollingBackoff : this.RunPollingInterval;
    }
}

File: 01-core-implementations/python/semantic_kernel/agents/open_ai/run_polling_options.py
#!/usr/bin/env python3
"""
Run Polling Options module

Copyright (c) 2025 Bryan Roe
Licensed under the MIT License

This file is part of the Semantic Kernel - Advanced AI Development Framework.
Original work by Bryan Roe.

Author: Bryan Roe
Created: 2025
License: MIT
"""

# Copyright (c) Microsoft. All rights reserved.

from datetime import timedelta

from pydantic import Field

from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.feature_stage_decorator import experimental

@experimental
class RunPollingOptions(KernelBaseModel):
    """Configuration and defaults associated with polling behavior for Assistant API requests."""

    default_polling_interval: timedelta = Field(default=timedelta(milliseconds=250))
    default_polling_backoff: timedelta = Field(default=timedelta(seconds=1))
    default_polling_backoff_threshold: int = Field(default=2)
    default_message_synchronization_delay: timedelta = Field(
        default=timedelta(milliseconds=250)
    )
    run_polling_interval: timedelta = Field(default=timedelta(milliseconds=250))
    run_polling_backoff: timedelta = Field(default=timedelta(seconds=1))
    run_polling_backoff_threshold: int = Field(default=2)
    message_synchronization_delay: timedelta = Field(
        default=timedelta(milliseconds=250)
    )
    message_synchronization_delay: timedelta = Field(
        default=timedelta(milliseconds=250)
    )
    message_synchronization_delay: timedelta = Field(default=timedelta(milliseconds=250))
    run_polling_timeout: timedelta = Field(default=timedelta(minutes=1))  # New timeout attribute

    def get_polling_interval(self, iteration_count: int) -> timedelta:
        """Get the polling interval for the given iteration count."""
        return (
            self.run_polling_backoff
            if iteration_count > self.run_polling_backoff_threshold
            else self.run_polling_interval
        )
